compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToUKernels.cpp

```diff
@@ -0,0 +1,263 @@
+struct FnNameAndDefAttrs {
+  std::string name;
+  SmallVector<NamedAttribute> defAttrs;
+};
+
+/// Returns the function name and attributes to use for a ukernel with given
+/// `ukernelName` on the target described by `targetAttr`.
+static FnNameAndDefAttrs getFnNameAndDefAttrs(
+    const char *ukernelName, RewriterBase &rewriter,
+    IREE::HAL::ExecutableTargetAttr targetAttr) {
+  FnNameAndDefAttrs result;
+  result.name = std::string("iree_uk_") + ukernelName;
+  result.defAttrs.emplace_back(
+      rewriter.getStringAttr("hal.import.fields"),
+      rewriter.getArrayAttr({rewriter.getStringAttr("processor_data")}));
+  result.defAttrs.emplace_back(rewriter.getStringAttr("hal.import.bitcode"),
+                               rewriter.getBoolAttr(true));
+  result.defAttrs.emplace_back(
+      rewriter.getStringAttr("hal.import.cconv"),
+      IREE::HAL::CallingConventionAttr::get(
+          rewriter.getContext(),
+          IREE::HAL::CallingConvention::ParameterStruct));
+  return result;
+}
+
+/// ============================= END ====================================
+/// ================== SAME UTILITIES AS IREE LLVMCPU ====================
+/// ======================================================================
+
+/// Matches a linalg.mmt4d operation sequence and converts it into a
+/// iree_codegen.ukernel."iree_uk_mmt4d" operation, that is later lowered into
+/// a call to the microkernel.
+static FailureOr<IREE::Codegen::UKernelOpInterface> matchDAGForUKernel(
+    RewriterBase &rewriter, linalg::Mmt4DOp op) {
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(op);
+  const char ukernelName[] = "mmt4d";
+  if (!hasUkernel(targetAttr, ukernelName)) {
+    return failure();
+  }
+  Value lhs = getInputForUKernel(op.getDpsInputOperand(0)->get());
+  Value rhs = getInputForUKernel(op.getDpsInputOperand(1)->get());
+  Value out = op.getDpsInitOperand(0)->get();
+  auto outType = llvm::cast<ShapedType>(out.getType());
+  Type lhsElemType = getElementTypeForUKernel(op.getDpsInputOperand(0)->get());
+  Type rhsElemType = getElementTypeForUKernel(op.getDpsInputOperand(1)->get());
+  Type outElemType = outType.getElementType();
+  uint32_t flags = 0;
+  if (lhsElemType.isBF16() && rhsElemType.isBF16() && outElemType.isF32()) {
+    flags = IREE_UK_FLAG_MMT4D_TYPE_BF16BF16F32;
+  } else if (lhsElemType.isBF16() && rhsElemType.isBF16() &&
+             outElemType.isBF16()) {
+    flags = IREE_UK_FLAG_MMT4D_TYPE_BF16BF16BF16;
+  } else {
+    return rewriter.notifyMatchFailure(
+        op, "unsupported combination of element types");
+  }
+
+  // // Check if the accumulator is zero-filled.
+  if (isInitializedToZero(out)) {
+    // Not setting flags |= IREE_UK_FLAG_MMT4D_ACCUMULATE, so the mmt4d op won't
+    // read the existing accumulator, so its defining op can be discarded.
+    if (auto fillOp = out.getDefiningOp<linalg::FillOp>()) {
+      out = fillOp.getDpsInitOperand(0)->get();
+    }
+  } else {
```

Comment:
It could be initialized to 1, or something else? I guess that would be fine though, it'd just keep that initialization in the IR (`out` unchanged)

Reply:
I don't think so. I might be wrong but I believe this just checks if the `fill` op is present in the function body (to consume it directly while lowering and not having to read the zero filled accumulator).

Soon all these aspects of the code will get clearer as I make incremental progress - currently that's my honest understanding of the state (and demonstrating the same via the lit tests).